draft-ietf-quic-transport.md

```diff
@@ -3018,11 +3018,19 @@ elements on the network path, an endpoint verifies the following when an ACK
 frame is received:
 
 * The increase in ECT(0) and ECT(1) counters MUST be at least the number of QUIC
-  packets newly acknowledged that were sent with the corresponding codepoint.
+  packets newly acknowledged that were sent with the corresponding codepoint
+  minus the increase in the CE counter. Detects network remarking
+  between ECT(0) and ECT(1).
 
 * The total increase in ECT(0), ECT(1), and CE counters reported in the ACK
   frame MUST be at least the total number of QUIC packets newly acknowledged in
-  this ACK frame.
+  this ACK frame. Detects if the network remark ECT(0), ECT(1) or CE to
+  Not-ECT.
+
+If a sender receives an ACK that contains no new acknowledgments, for example
+due to reordering of the ACKs, then ECN counter comparison SHOULD NOT be
+performed. Also if sender do not have state to determine if a particular PSN
+is newly acknowledge or not, then the comparison SHOULD NOT be performed.
```

Comment:
The implication here seems to be that an implementation should retain extra state to distinguish between a packet that is deemed lost but not yet acknowledged and a packet that was previously acknowledged. Is that intended?

If these two cases are instead conflated, e.g. by forgetting about sent packets once either acknowledged or deemed lost, then at worst the number of newly acknowledged packets in an ACK would sometimes be under-counted, and a verification with a higher rate of false positives seems preferable to going to extra effort to decide to skip verification outright.

Reply:
Partly, sender need to track if a packet has been acknowledged or not. But, my assumption is that the sender will have a window where such information will be discarded when likely not relevant anymore. So if everything has been ACKed up to PSN=48. Then cleaning up your state and discarding all below 48 is fine. When the reorderd ACK that indicate reception of PSN 44 and 45 can then be determined to be behind this window and ignored. The second sentence is trying to state that for this example case this is fine to do.